components/nimbus/src/tests/stateful/helpers.rs

```diff
@@ -0,0 +1,98 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+use crate::{
+    glean_metrics::nimbus_events::{enrollment_status, EnrollmentStatusExtra},
+    metrics::{EnrollmentStatusExtraDef, MetricsHandler},
+};
+use glean::RecordedEvent;
+use std::sync::{Mutex, MutexGuard};
+
+impl From<EnrollmentStatusExtraDef> for EnrollmentStatusExtra {
+    fn from(enrollment_status_extra: EnrollmentStatusExtraDef) -> Self {
+        EnrollmentStatusExtra {
+            branch: enrollment_status_extra.branch,
+            conflict_slug: enrollment_status_extra.conflict_slug,
+            error_string: enrollment_status_extra.error_string,
+            reason: enrollment_status_extra.reason,
+            slug: enrollment_status_extra.slug,
+            status: enrollment_status_extra.status,
+        }
+    }
+}
+
+/// Implemented on RecordedEvents to aid in verifying the value of keys on event extras
+pub trait GetExtraKey {
+    fn get_extra_key(&self, key: &str) -> String;
+}
+
+impl GetExtraKey for RecordedEvent {
+    fn get_extra_key(&self, key: &str) -> String {
+        self.extra.clone().unwrap().get(key).unwrap().to_owned()
+    }
+}
+
+static METRICS_SINGLE_THREAD_MUTEX: Mutex<()> = Mutex::new(());
+
+/// A Rust implementation of the MetricsHandler trait
+/// Used to test recording of Glean metrics across the FFI within Rust
+///
+/// *NOTE: Use this struct's `new` method when instantiating it to lock the Glean store*
+pub struct TestMetrics<'a> {
+    _lock: Option<MutexGuard<'a, ()>>,
+}
+
+impl<'a> TestMetrics<'a> {
+    /// This struct does two things on initialization:
+    /// 1. It resets Glean
+    /// 2. It locks a static Mutex and hold onto its guard
+    ///
+    /// The purpose of the latter is so it can be used as such at the beginning of a test:
+    /// ```rust
+    /// let metrics = TestMetrics::new();
+    /// ```
+    /// and it will lock the Mutex so no other tests that use Glean can run at the same time.
+    ///
+    /// *NOTE: Cloned copies of TestMetrics objects do not lock the Mutex*
+    pub fn new() -> Self {
+        let lock = match METRICS_SINGLE_THREAD_MUTEX.lock() {
+            Ok(l) => l,
+            Err(poisoned) => poisoned.into_inner(),
+        };
```

Comment:
Okay so it's _kinda_ less cursed now I think? At least now whenever we use the `TestMetrics` handler in a test we know that it'll run while no other tests that use Glean run.